### PR TITLE
SONARPY-1519 Add python 3.12 to PythonVersionUtils

### PR DESCRIPTION
**Pull request by jdodinh**

### CODE REVIEW

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 1]
```diff
@@ -214,6 +214,9 @@ public static boolean isValidForProjectPythonVersion(List<String> validForPython
     if (validForPythonVersions.isEmpty()) {
       return true;
     }
+    if (supportedPythonVersions.stream().allMatch("312"::equals) && validForPythonVersions.contains("311")) {
+      return true;
```
Comment (guillaume-dequenne-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/TypeShed.java:218]: I had the impression you pushed this, but it seems not: there should be a `TODO` comment here with link to SONARPY-1522.

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
Looks good!
Left some minor comments. I did create SONARPY-1523 to add an integration test about this as well - I think it's okay to skip it for this PR.

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java (hunk 1 above)
Comment (jdodinh) [reply python-frontend/src/main/java/org/sonar/python/types/TypeShed.java:220]: Addressed in 5b320b2eae5bfcb712c343ad32d7b891c30bdb1b.

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 2]
```diff
@@ -214,6 +214,9 @@ public static boolean isValidForProjectPythonVersion(List<String> validForPython
     if (validForPythonVersions.isEmpty()) {
       return true;
     }
+    if (supportedPythonVersions.stream().allMatch("312"::equals) && validForPythonVersions.contains("311")) {
```
[the remark this answers is not in the document]
Comment (jdodinh) [reply python-frontend/src/main/java/org/sonar/python/types/TypeShed.java]: Addressed in 5b320b2eae5bfcb712c343ad32d7b891c30bdb1b.

File: python-frontend/src/test/java/org/sonar/python/types/TypeShedTest.java [hunk 3]
```diff
@@ -440,7 +440,12 @@ void pythonVersions() {
     ClassSymbol intSymbol = TypeShed.typeShedClass("int");
     assertThat(intSymbol.resolveMember("bit_count")).isNotEmpty();
 
+    setPythonVersions(PythonVersionUtils.fromString("3.12"));
+    intSymbol = TypeShed.typeShedClass("int");
+    assertThat(intSymbol.resolveMember("bit_count")).isNotEmpty();
+
     setPythonVersions(PythonVersionUtils.allVersions());
+
```
[the remark this answers is not in the document]
Comment (jdodinh) [reply python-frontend/src/test/java/org/sonar/python/types/TypeShedTest.java]: Addressed in 5b320b2eae5bfcb712c343ad32d7b891c30bdb1b.

File: python-frontend/src/main/java/org/sonar/python/types/TypeShed.java [hunk 4]
```diff
@@ -214,7 +214,9 @@ public static boolean isValidForProjectPythonVersion(List<String> validForPython
     if (validForPythonVersions.isEmpty()) {
       return true;
     }
-    if (supportedPythonVersions.stream().allMatch("312"::equals) && validForPythonVersions.contains("311")) {
+    // TODO: SONARPY-1522 - remove this workaround when we will have all the stubs for Python 3.12.
+    if (supportedPythonVersions.stream().allMatch(PythonVersionUtils.Version.V_312.serializedValue()::equals)
+      && validForPythonVersions.contains(PythonVersionUtils.Version.V_312.serializedValue())) {
```
Comment (jdodinh) [inline python-frontend/src/main/java/org/sonar/python/types/TypeShed.java]: Changed the bug in version number to `PythonVersionUtils.Version.V_312` in 159351aadb39f43eaca139cc83b23b716b91769a.

**Review by guillaume-dequenne-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**